src/h3/mod.rs

```diff
@@ -1174,15 +960,40 @@ impl Connection {
         Ok(())
     }
 
-    fn handle_stream(
+    fn process_control_stream(
         &mut self, conn: &mut super::Connection, stream_id: u64,
     ) -> Result<()> {
-        let stream = self
-            .streams
+        match self.process_readable_stream(conn, stream_id) {
+            Ok(_) => (),
+
+            Err(Error::Done) => (),
+
+            Err(e) => return Err(e),
+        };
+
+        if conn.stream_finished(stream_id) {
+            conn.close(
+                true,
+                Error::ClosedCriticalStream.to_wire(),
+                b"Critical stream closed.",
+            )?;
+
+            return Err(Error::ClosedCriticalStream);
+        }
+
+        Ok(())
+    }
+
+    fn process_readable_stream(
+        &mut self, conn: &mut super::Connection, stream_id: u64,
+    ) -> Result<(u64, Event)> {
+        self.streams
             .entry(stream_id)
             .or_insert_with(|| stream::Stream::new(stream_id, false));
 
         loop {
+            let stream = self.streams.get_mut(&stream_id).unwrap();
```

Comment:
`unwrap()` in library code. But also I don't understand why the assignment of `stream` was moved inside the loop. Since stream_id doesn't change inside the function, it seems like a wasted `get_mut()` call each iteration, and it makes the previous `.entry().or_insert_with()` weird.

Reply:
Removed the unwrap, thanks.

As for the `get_mut()` thing, this is necessary to avoid holding up `self` in a borrow for the entire duration of the loop, otherwise we wouldn't be able to borrow it again when calling `process_frame()` below (this is due to a borrow checker limitation which prevents it from working across function calls).

```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
    --> src/h3/mod.rs:1186:27
     |
999  |             let stream = match self.streams.get_mut(&stream_id) {
     |                                ------------ first mutable borrow occurs here
...
1006 |             match stream.state() {
     |                   ------ first borrow later used here
...
1186 |                     match self.process_frame(conn, stream_id, frame) {
     |                           ^^^^ second mutable borrow occurs here

error: aborting due to previous error
```

I added a comment to make this somewhat more clear.